### PR TITLE
Robust Trees initial commit

### PR DESCRIPTION
**Pull request by nathanljustin**

### CODE REVIEW

File: RobustTrees/RobustOCT.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+        if isinstance(self.y, [pd.Series, pd.DataFrame]):
+            self.y = y.values
+            self.labels = np.unique(self.y)
+        else:
+            self.y = y
+            self.labels = np.unique(self.y)
+        # datapoints contains the indicies of our training data
+        self.datapoints = np.arange(0, self.X.shape[0])
+        self.tree = tree
```
Comment (patrickvossler18) [inline RobustTrees/RobustOCT.py:38]: Sorry, I forgot to push my changes for the Strong Trees in which we handle the data conversion before initializing the Gurobi model. The reason for this is because we need to keep the column names for any tree visualizations, etc but if we keep the data in a dataframe, we lose access to the sklearn API

See below for how to restructure this part to match Strong Trees:
* We now use an [extract metadata function](https://github.com/patrickvossler18/decision_tree_estimators/blob/99418904179a87b38ebb565e60be0a706f2b2025/StrongTrees/StrongTree.py#L50) to store column names and label information
* We pass [the column information](https://github.com/patrickvossler18/decision_tree_estimators/blob/99418904179a87b38ebb565e60be0a706f2b2025/StrongTrees/FlowOCT.py#L9) to the RobustOCT class

Let me know if you have any questions about these changes

File: RobustTrees/RobustTreeUtils.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    label_i = master.y[i]
+    target = [] # list of nodes whose edge to the sink is part of the min cut
+
+    # Solve shortest path problem via DFS w/ pruning
+    target, cost, xi, v = shortest_path_solver(master, i, label_i, terminal_nodes, terminal_path_dict, terminal_features_dict, 
+        terminal_assignments_dict, terminal_cutoffs_dict, initial_xi, initial_mins, initial_maxes)
+    
+    return target, xi, cost, v
+
+def mycallback(model, where):
```
Comment (patrickvossler18) [inline RobustTrees/RobustTreeUtils.py:241]: Do we need this callback? My understanding is that this was mostly used for the paper replication code

**Review by patrickvossler18 — CHANGES_REQUESTED** [review]
Hi Nathan, thank you for pulling together the robust trees code. I have a few questions and there are some small changes that I have commented on in my review. Feel free to message me if anything I wrote was unclear!

File: RobustTrees/RobustOCT.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        # Gurobi model
+        self.model = Model('RobustOCT')
+        # The cuts we add in the callback function would be treated as lazy constraints
+        self.model.params.LazyConstraints = 1
+
+        '''
+        To compare all approaches in a fair setting we limit the solver to use only one thread to merely evaluate 
+        the strength of the formulation.
+        '''
+        # self.model.params.Threads = 1
```
[the remark this answers is not in the document]
Comment (nathanljustin) [reply RobustTrees/RobustOCT.py:114]: Yes. From the [Gurobi reference manual](https://www.gurobi.com/wp-content/plugins/hd_documentations/documentation/9.0/refman.pdf):
> It will generally use all of the cores in the machine, but it may choose to use fewer.

 We limited it to 1 thread to have consistency of runtimes between instances for comparison, but I assume that we would want users to be able to set this themselves or use the default.

File: RobustTrees/RobustOCT.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        self.model.params.TimeLimit = time_limit
+        
+        # NOTE - Should we allow users to suppress logging?
+        # self.model.params.OutputFlag = 0 # supress all logging
+
+        '''
+        The following variables are used for the Benders problem to keep track of the times we call the callback.
```
[the remark this answers is not in the document]
Comment (nathanljustin) [reply RobustTrees/RobustOCT.py:121]: So the Robust Trees formulation by default can only be solved using a Benders decomposition approach (unlike the strong trees approach). These variables keep track of the all the callback information, however I think some of them are unnecessary (i.e. they say the same thing as some other variables) and so I can safely remove some of them.

File: RobustTrees/RobustTreeUtils.py (hunk 2 above)
Comment (nathanljustin) [reply RobustTrees/RobustTreeUtils.py:241]: Yes, so we do need the callback to feed to the gurobi solver in the fitting step since Robust Trees can only be solved using a Benders approach with callbacks

File: trees/RobustTree.py [hunk 5]
```diff
@@ -1,105 +1,15 @@
...
-        return prediction
 
+# Include Necessary imports in the same folder
+from trees.utils.Tree import Tree
+from trees.utils.RobustOCT import RobustOCT
+from trees.utils.RobustTreeUtils import mycallback, check_integer, check_same_as_X
+import time
 
 class RobustTreeClassifier(ClassifierMixin, BaseEstimator):
     """ An example classifier which implements a 1-NN algorithm.
```
Comment (patrickvossler18) [inline trees/RobustTree.py:15]: We need to add a short description of the class here because it will be used to create the API documentation

File: trees/utils/RobustOCT.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+        '''
+        To compare all approaches in a fair setting we limit the solver to use only one thread to merely evaluate 
+        the strength of the formulation.
+        '''
+        # self.model.params.Threads = 1
+        self.model.params.TimeLimit = time_limit
+        
+        # NOTE - Should we allow users to suppress logging?
+        # self.model.params.OutputFlag = 0 # supress all logging
```
Comment (patrickvossler18) [inline trees/utils/RobustOCT.py:76]: Yes, I think we should give users this option. Maybe with a verbose argument for when they initialize the class

**Review by patrickvossler18 — APPROVED** [review]
Looks good! I'm going to merge these changes in but the remaining to-dos are the following:

- [ ] Add description of RobustTreeClassifier in the docstring
- [ ] Add argument for option to suppress logging when fitting tree
